_posts/2018-10-20-reducers-exploring-state-management-in-react.md

```diff
@@ -0,0 +1,125 @@
+
+Let's examine the code snippet below.
+
+```jsx
+function reduce(prevState, { type, payload }) {
+  switch (type) {
+    case "ADD":
+      return {
+        items: [...prevState.items, prevState.newItem],
+        newItem: ""
+      };
+    case "REMOVE":
+      return {
+        items: prevState.items.filter(
+          (i, idx, prevItems) => idx !== prevItems.indexOf(payload)
+        )
+      };
+    case "CHANGE":
+      return { newItem: payload };
+    case "DISCOUNTS":
+      return { discounts: payload };
+    case "INITIAL":
+      return {
+        newItem: "",
+        items: [],
+        discounts: []
+      };
+    default:
+      throw new Error("Unknown message type received");
+  }
+}
+
+class Cart extends React.Component {
+  state = reduce(undefined, { type: "INITIAL" });
+
+  componentDidUpdate(prevProps, prevState) {
+    if (prevState.items.length !== this.state.items.length) {
+      const itemsQuery = this.state.items.join(",");
+
+      ajax(`https://discountdb.com/?items=${itemsQuery}`)
+        .then(discounts => 
+          this.dispatch({ type: "DISCOUNTS", payload: discounts }));
+    }
+  }
+
+  dispatch = action => 
+    this.setState(prevState => reduce(prevState, action));
+
+  onNewItemChange = event =>
+    this.dispatch({ type: "CHANGE", payload: event.target.value });
+
+  addItem = () => this.dispatch({ type: "ADD" });
+
+  removeItem = item => 
+    () => this.dispatch({ type: "REMOVE", payload: item });
+
+  render() {
+    return (
+      <ul>
+        <h2>Cart</h2>
+        <input
+          type="text"
+          onChange={this.onNewItemChange}
+          value={this.state.newItem}
+        />
+        <button onClick={this.addItem}>Add item</button>
+
+        {this.state.items.map((item, idx) => (
+          <li key={idx}>
+            <button onClick={this.removeItem(item)}>Remove</button>
+            {item}, &ensp;
+            {this.state.discounts[idx]}% off!
+          </li>
+        ))}
+      </ul>
+    );
+  }
+}
+```
+
+[![Edit zrww3wp57m](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/zrww3wp57m)
+
+Here our `reduce` function resolves to a switch statement which delegates according to certain messages. We call this function in two places; we call it directly to initialize our component state and we call it in our `dispatch` function to allow our event handlers to dynamically pass messages.
+
+Unlike the function you pass to `Array.prototype.reduce`, our reduce only returns the changes to state and not the entire new state. This is because we only pass changes to `this.setState`.
+
+_Notice that the only place we call `this.setState` directly is in the `dispatch` function._
+
+## Benefits
+
+What we have done is extract and enumerate the various transformations that can happen to our component state.
```

Comment:
💡 It feels like after all that explanation, the benefits and drawbacks aren't given much space.

Have you used this technique yourself? How did it go?

Reply:
I agree with your point, and think your questions are good. 

I worked this technique into the react app for my [contact](https://github.com/mhanberg/contact) project and it gave me some insight.

_and also reminded me how terrible that React code is..._